Update config

johnbdesanto opened the pull request:
This PR is meant to address issue #356 and makes the following changes:

  -Adds generic "format" option to the InputData field used throughout the configuration file. This is currently optional and not used anywhere in the code, but allows us the ability to develop tools for reading inputs from multiple file types in the future.
  -Adds a test to make sure that the format exists when reading an example InputData field.
  -Moves the travel time input from "main" to "posfilter" in the config file. Removes references to loading input data from "main" and instead only loads data from "posfilter" or "solver", if those options are set.
  -Updates documentation to reflect these changes.

Note that currently the pytest infrastructure is failing due to the removal of the input data from "main". The issue appears to stem from an empty _all_files_dict_ array, meaning that no input files from any mode are being recognized. However, an independent test run processing data in a jupyter notebook with the new configuration file structure was successful. As it stands, feedback on the potential issue here and a way to test this changes is required.

johnbdesanto commented:
@lsetiawan and I identified and solved the bug that was leading to the broken tests, which was an if statement trying to compare a string to a boolean. I have gone ahead and cleaned up this bug and now the tests are all passing, so I have marked this PR as ready for review.

lsetiawan commented on src/gnatss/configs/main.py:
```diff
@@ -39,10 +39,10 @@ class ArrayCenter(BaseModel):
-    travel_times: InputData | None = Field(
-        None, description="Input travel times data path specification"
-    )
+# class MainInputs(BaseModel):
+#     travel_times: InputData | None = Field(
+#         None, description="Input travel times data path specification"
+#     )
```
I would just remove all the unnecessary code rather than commenting

lsetiawan replied:
```suggestion
```

lsetiawan commented on src/gnatss/ops/io.py:
```diff
@@ -184,7 +182,7 @@ def gather_files_all_procs(
-        if proc == "main" or hasattr(config, proc):
+        if hasattr(config, proc):
             all_files_dict.update(gather_files(config, proc, mode))
     return all_files_dict
```
For the sake of non silent failure, maybe raise error above this like if all_files_dict is empty

johnbdesanto replied:
I added a ValueError message in case all_files_dict{}, see lines 188-190 of ops/io.py. However, after doing some testing I could not get this error to trigger when running an example in a Jupyter notebook. I tried various flavors of commenting out input data fields in the config yaml and all of my permutation failed at the data_loading() stage upstream of gather_files_all_procs().

While this may be a good thing overall from the perspective of future-proofing, I can't say with certainty whether this will ever come into play (although I guess there is precedent).

lsetiawan commented on src/gnatss/ops/io.py:
```diff
@@ -184,8 +182,13 @@ def gather_files_all_procs(
+        if hasattr(config, proc):
             all_files_dict.update(gather_files(config, proc, mode))
+
+    if not all_files_dict:
+        msg = "Error: gather_files_all_procs() failed to parse input data."
+        raise ValueError(msg)
```
```suggestion
    if not all_files_dict:
        avail_procs = ', '.join(constants.DEFAULT_CONFIG_PROCS)
        msg = "Could not find valid process in configuration. Valid options: {avail_procs}."
        raise ValueError(msg)
```

This is great, however I'd make the error message a little more clear and explicit. I'd suggest adding a test for this. To trigger it, simply remove the two processing attributes in the input config object, that should trigger this error. I know it's a little bit of work, but I think it's a good exercise to creating test and really thinking about what is happening within this section of the software 😄 Let me know if you have any questions!

johnbdesanto replied:
I'll see what I can put together in the next day or so and then reach back out.

lsetiawan commented on tests/conftest.py:
```diff
@@ -41,7 +41,7 @@ def blank_env(blank_csv_test_file: Path) -> None:
         "GNATSS_ARRAY_CENTER__LON": str(2.2),
         "GNATSS_TRANSPONDERS": "[]",
-        "GNATSS_INPUT_FILES": dumps({"travel_times": {"path": str(blank_csv_test_file)}}),
+        # "GNATSS_INPUT_FILES": dumps({"travel_times": {"path": str(blank_csv_test_file)}}),
```
```suggestion
```

johnbdesanto commented on tests/test_loaders.py:
```diff
@@ -62,6 +63,17 @@ def test_load_configuration_valid_path(config_yaml_path):
 
+@pytest.mark.parametrize(
+    "config_yaml_path",
+    [TEST_DATA_FOLDER / "config.yaml"],
```
Done

johnbdesanto commented:
I found a bug introduced by this PR in further testing. The issue was that the preprocess_data() subroutine was called in main(), and would fail if no travel time data were read as input if you ran --solver or --from-cache. This was previously not a problem since travel time input was required as a main input file in the config, but as of this update travel times are now accepted as posfilter input and therefore optional.

The fix I implemented was to move the location where preprocess_data() is called from main() to run_posfilter() since it is no longer needed outside the posfilter. In addition, there was a subroutine called by preproces_data() to calculate the harmonic mean sound velocity; since this functionality is only needed when the solver is called I have moved this to run_solver().

@lsetiawan Please review this change.
